Add ability to sniff media content should we failed to play a file or if the provided content type was invalid.

jyavenard opened the pull request:
#### 93f6303ff83e5c20dbd79f5e6ae7c8a801ee519d
<pre>
Add ability to sniff media content should we failed to play a file or if the provided content type was invalid.
<a href="https://bugs.webkit.org/show_bug.cgi?id=270975">https://bugs.webkit.org/show_bug.cgi?id=270975</a>
<a href="https://rdar.apple.com/problem/124614908">rdar://problem/124614908</a>

Reviewed by Jer Noble.

We have always relied on the server to provide a valid content-type and if that failed
we always used the URL&apos;s file name extension instead.
While the HTML5 specs clearly states that determining the type of a resource
should be done through sniffing, this is however a too significant change
to enable right away. There are also performance advantages in using the provided
content-type: it&apos;s typically immediately available.
Should the provided content-type be invalid or non-existent (such as with some blobs)
playback would have failed.

We add a MIMEtype sniffer for media content as per HTML5 specs.
Should we fail to play a media using the older content type detection rather
than immediately fail, we will also sniff the type as a last attempt before
retrying.

In order to not unnecessarily retry following a sniffing, we needed to
distinguish a format error from a network error which the MediaPlayerPrivateAVFoundationObjC
didn&apos;t do. As a fly-by fix, we add a way to distinguish the two errors by
recording in the WebCoreNSURLSession any network failures.

For now will limit sniffing to media element where the src attribute is set
(and so exclude element where alternative sources are defined)

Added API tests for the mimetype sniffer and tests for webm and mp4 in a blob.

* LayoutTests/media/video-src-mp4-blob-expected.txt: Added.
* LayoutTests/media/video-src-mp4-blob.html: Added.
* LayoutTests/media/video-src-webm-blob-expected.txt: Added.
* LayoutTests/media/video-src-webm-blob.html: Added.
* LayoutTests/media/video-srcobject-mp4-blob-expected.txt: Added.
* LayoutTests/media/video-srcobject-mp4-blob.html: Added.
* LayoutTests/http/tests/media/video-throttled-load-metadata-expected.txt:
* LayoutTests/http/tests/media/video-throttled-load-metadata.html: The test was racy and could
caused the loadedmetadata message to have been received before the worker&apos;s message.
* LayoutTests/platform/mac-wk1/TestExpectations:
* LayoutTests/platform/mac/TestExpectations: Remove failure expectations. Tests were failing as an incorrect
content-type was provided. Now that we sniff if the provided content-type was incorrect
files can properly play again.
* Source/WTF/wtf/NativePromise.h:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/html/HTMLMediaElement.cpp: Move all the logic from MediaPlayer related
to guessing the ContentType if missing into the HTMLMediaElement.
If all have failed, we will sniff the content.
(WebCore::HTMLMediaElement::~HTMLMediaElement):
(WebCore::HTMLMediaElement::loadResource): If the content is blocked, signal it as a network error rather than content error.
Otherwise we will attempt to load the content twice (once for sniffing and once for playing) causing test failures.
(WebCore::HTMLMediaElement::needsContentTypeToPlay const): Add methods to determine if sniffing is ultimately required.
(WebCore::HTMLMediaElement::sniffForContentType):
(WebCore::HTMLMediaElement::mediaLoadingFailed): Add one ultimate attempt if we failed to decode and
didn&apos;t encounter a network error and didn&apos;t attempt to previously sniff content.
(WebCore::HTMLMediaElement::cancelPendingTasks):
(WebCore::HTMLMediaElement::cancelSniffer):
(WebCore::HTMLMediaElement::mediaPlayerEngineFailedToLoad): Save if a network error was encountered.
If so, we won&apos;t attempt to sniff later on. This is required as multiple tests are
relying on the network being accessed only once, doing one extra access for sniffing
made the tests fail.
(WebCore::HTMLMediaElement::mediaPlayerEngineFailedToLoad const): Deleted.
* Source/WebCore/html/HTMLMediaElement.h:
* Source/WebCore/platform/CommonAtomStrings.h:
* Source/WebCore/platform/ContentType.cpp:
(WebCore::ContentType::ContentType):
(WebCore::ContentType::fromURL):
* Source/WebCore/platform/ContentType.h:
(WebCore::ContentType::typeWasInferredFromExtension const):
* Source/WebCore/platform/PlatformMediaError.cpp:
(WebCore::convertEnumerationToString):
* Source/WebCore/platform/PlatformMediaError.h:
* Source/WebCore/platform/graphics/MIMESniffer.cpp: Added.
(WebCore::MIMESniffer::span8):
(WebCore::MIMESniffer::hasSignatureForMP4):
(WebCore::MIMESniffer::parseWebMVint):
(WebCore::MIMESniffer::hasSignatureForWebM):
(WebCore::MIMESniffer::matchMP3Header):
(WebCore::MIMESniffer::mp3FrameSize):
(WebCore::MIMESniffer::parseMP3Frame):
(WebCore::MIMESniffer::hasSignatureForMP3WithoutID3):
(WebCore::MIMESniffer::mimeTypeFromSnifferEntries):
(WebCore::MIMESniffer::getMIMETypeFromContent):
* Source/WebCore/platform/graphics/MIMESniffer.h: Added.
* Source/WebCore/platform/graphics/MediaPlayer.cpp:
(WebCore::applicationOctetStream):
(WebCore::MediaPlayer::load):
(WebCore::MediaPlayer::loadWithNextMediaEngine):
* Source/WebCore/platform/graphics/MediaPlayer.h:
(WebCore::MediaPlayerClient::mediaPlayerEngineFailedToLoad):
(WebCore::MediaPlayerClient::mediaPlayerEngineFailedToLoad const): Deleted.
* Source/WebCore/platform/graphics/MediaResourceSniffer.cpp: Added.
(WebCore::MediaResourceSniffer::create):
(WebCore::MediaResourceSniffer::MediaResourceSniffer):
(WebCore::MediaResourceSniffer::~MediaResourceSniffer):
(WebCore::MediaResourceSniffer::cancel):
(WebCore::MediaResourceSniffer::promise const):
(WebCore::MediaResourceSniffer::dataReceived):
(WebCore::MediaResourceSniffer::loadFailed):
(WebCore::MediaResourceSniffer::loadFinished):
* Source/WebCore/platform/graphics/MediaResourceSniffer.h: Added.
* Source/WebCore/platform/graphics/avfoundation/MediaPlayerPrivateAVFoundation.cpp:
(WebCore::MediaPlayerPrivateAVFoundation::updateStates): We needed to be able
to distinguish FormatError vs NetworkError so that we don&apos;t unnecessarily retry
playback following sniffing.
* Source/WebCore/platform/graphics/avfoundation/MediaPlayerPrivateAVFoundation.h:
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateAVFoundationObjC.mm:
(WebCore::MediaPlayerPrivateAVFoundationObjC::assetStatus const): Distinguish
load failure due to decoding error vs networking error.
* Source/WebKit/GPUProcess/media/RemoteMediaPlayerProxy.cpp:
(WebKit::RemoteMediaPlayerProxy::mediaPlayerEngineFailedToLoad):
(WebKit::RemoteMediaPlayerProxy::mediaPlayerEngineFailedToLoad const): Deleted.
* Source/WebKit/GPUProcess/media/RemoteMediaPlayerProxy.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Tools/TestWebKitAPI/CMakeLists.txt:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebCore/MIMESniffer.cpp: Added.
(TestWebKitAPI::TEST):

Canonical link: <a href="https://commits.webkit.org/276258@main">https://commits.webkit.org/276258@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/85db953a3a8f96126cbdcdfc2d80c1697c1f6b24

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/44184 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/23253 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/46619 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/46828 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/40211 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/46489 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/27201 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/20647 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/36404 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/44762 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/20297 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/38060 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/17438 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/17780 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/39164 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/2234 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/37506 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/40373 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/39454 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/48429 "Built successfully") | 
| [  ~~🛠 🧪 jsc-arm64~~](https://ews-build.webkit.org/#/builders/12/builds/43742 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/19164 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/15738 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/43278 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/20525 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/38331 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/42003 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9816 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/20749 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/50824 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/20151 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/10279 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

jernoble commented on Source/WebCore/platform/graphics/MediaResourceSniffer.h:
```diff
@@ -0,0 +1,61 @@
+    : public PlatformMediaResourceClient {
+    WTF_MAKE_FAST_ALLOCATED;
+public:
+    static RefPtr<MediaResourceSniffer> create(PlatformMediaResourceLoader&, ResourceRequest&&, size_t maxSize = -1);
```
Nit: consider an `std::optional<size_t>` here, rather than a `-1`, or (ref: [Analyzing-Build-Performance](https://github.com/WebKit/WebKit/wiki/Analyzing-Build-Performance#avoid-default-parameters)), use overloading here to avoid the default parameter.

jernoble commented on Source/WebCore/html/HTMLMediaElement.cpp:
```diff
@@ -1678,55 +1681,105 @@ void HTMLMediaElement::loadResource(const URL& initialURL, const ContentType& co
+}
 
-    mediaPlayerRenderingModeChanged();
+void HTMLMediaElement::sniffForContentType(const URL& url, Function<bool(const ContentType&)>&& completionHandler)
```
I'm curious why this takes a `Function` rather than returning a `NativePromise`?

jyavenard replied:
it just made it easier in this context, the function may not be called : sniffForContentType already checks for the WeakPtr and that the sniffing operation hasn't been cancelled. I would have had to run those checks for every caller otherwise.
it also avoids an extra dispatch to stay closer to the existing implementation.

cdumez replied:
It is weird to have an async function that takes in a completion handler not consistently call its completion handler. I think this should take an actual `WTF::CompletionHandler` type is should be called consistently. It's really not a great API otherwise.

jyavenard replied:
point noted

jernoble commented on Source/WebCore/platform/graphics/MediaResourceSniffer.cpp:
```diff
@@ -0,0 +1,102 @@
+        return;
+    }
+    if (!m_producer)
+        m_producer.resolve(WTFMove(mimeType));
```
Ditto re: inverted condition.

jyavenard replied:
I added a Producer::isSettled() and will use that instead, it is indeed more explicit.

jernoble commented on Source/WebCore/platform/graphics/MediaResourceSniffer.cpp:
```diff
@@ -0,0 +1,102 @@
+    if (auto resource = WTFMove(m_resource))
+        resource->shutdown();
+    if (!m_producer)
+        m_producer.reject(PlatformMediaError::Cancelled);
```
Nit: This seems like an inverted condition. :)

Apparently `operator bool` for a Producer returns whether the contained Promise is settled, but this _looks_ super weird, because it implies that `m_producer` is a pointer (which it _kind of_ is because it also has a `operator ->`.

This could be done in another patch, but I think Producer's `operator bool` should be replaced with an explicit `isSettled()`.

jyavenard replied:
Producer::operator bool() is equivalent to promise::isSettled() ; I could is isNothing() instead.

jernoble commented on Source/WebCore/platform/graphics/MediaResourceSniffer.cpp:
```diff
@@ -0,0 +1,102 @@
+
+void MediaResourceSniffer::dataReceived(PlatformMediaResource&, const SharedBuffer& buffer)
+{
+    auto mimeType = MIMESniffer::getMIMETypeFromContent({ buffer.data(), buffer.size() });
```
This seems incorrect; it would only work on the first data that's received. Shouldn't the `m_content.append(buffer)` step happen first, and _then_ the sniff?

jyavenard replied:
good point, I believe that in practice it will always work on the first packet received, as it's unlikely we will receive less than the 1485 bytes we requested.

cdumez replied:
`{ buffer.data(), buffer.size() }` -> `buffer.dataAsSpanForContiguousData()`. It is a horrible name so I intend to rename soon but still.

jernoble commented on Source/WebCore/platform/ContentType.h:
```diff
@@ -28,12 +28,18 @@
     WEBCORE_EXPORT explicit ContentType(String&& type);
     WEBCORE_EXPORT explicit ContentType(const String& type);
+    WEBCORE_EXPORT ContentType(const String& type, const WTF::URL&);
+    WEBCORE_EXPORT ContentType(const String& type, bool);
```
Is this second constructor really needed? I don't see it used anywhere.

jyavenard replied:
it's used by the IPC serialiser. It is very weird how the serialiser works IMHO. I'll add a comment

cdumez replied:
You could make it private and mark the serializer class as friend like we often do.

jyavenard replied:
I now use it following the move to have a `fromURL()` method

jernoble commented on Source/WebCore/platform/ContentType.h:
```diff
@@ -28,12 +28,18 @@
 public:
     WEBCORE_EXPORT explicit ContentType(String&& type);
     WEBCORE_EXPORT explicit ContentType(const String& type);
+    WEBCORE_EXPORT ContentType(const String& type, const WTF::URL&);
```
Since the URL isn't stored, I think this makes more sense as a static method; something like:
```
    WEBCORE_EXPORT static fromURL(const WTF::URL&);
```

And the call site would change from:

```
        // Attempt to determine the content-type from the URL file extension.
        contentType = ContentType { emptyString(), url };
```

To:

```
        // Attempt to determine the content-type from the URL file extension.
        contentType = ContentType::fromURL(url);
```

jyavenard replied:
good idea.

jernoble commented on Source/WebCore/platform/graphics/MIMESniffer.cpp:
```diff
@@ -0,0 +1,341 @@
+
+    // This ignores the four bytes that correspond to the version number of the "major brand".
+    // While bytes-read is less than box-size, continuously loop through these steps:
+    while (bytesRead < boxSize) {
+        // If the three bytes from sequence[bytes-read] to sequence[bytes-read + 2] are equal to 0x6D 0x70 0x34 ("mp4"), return true.
+        if (sequence[bytesRead] == 0x6d && sequence[bytesRead + 1] == 0x70 && sequence[bytesRead + 2] == 0x34)
+            return true;
+        // Increment bytes-read by 4.
+        bytesRead += 4;
+    }
```
Can we leverage ISOBox::peekBox() here?

jyavenard replied:
I wasn't aware of that method, however I really wanted us to be 100% spec compliant too and considering all the steps on how to read the box are defined in the spec; I feel it's better to follow it there.

cdumez commented on Source/WebCore/html/HTMLMediaElement.cpp:
```diff
@@ -1678,55 +1681,105 @@ void HTMLMediaElement::loadResource(const URL& initialURL, const ContentType& co
+    if (!m_sniffer)
+        return handler(ContentType { });
+
+    m_sniffer->promise()->whenSettled(RunLoop::main(), [weakThis = WeakPtr { *this }, completionHandler = WTFMove(handler)](auto&& result) mutable {
```
It doesn't seem like this lambda needs to capture `weakThis` at all. I think it is the `completionHandler` that should capture `weakThis` (and it actually currently captures `this` and uses it) and do the null check.

Otherwise, the completion handler capturing `this` in the function above looks unsafe and is only not crashing because of a weakPtr check at the call site, which is too fragile.

jyavenard replied:
That was the purpose of taking the weakPtr ; `sniffForContentType` is the one ensuring that the completionHandler is only called if `this` is valid and that the operation wasn't cancelled.

It may "look" unsafe ; but it really isn't

jyavenard replied:
I could change the name of `sniffForContentType` to indicate; like what we've done in other places like `ensureWeakOnHTMLMediaElement`

cdumez replied:
As I commented elsewhere, I think `completionHandler` should always be called so 🤷🏻

jyavenard replied:
ok, I'll stop using a completionHandler then and use a full promise path, so that it can be rejected.

cdumez replied:
CompletionHandler could be called with a std::expected or std::optional or ExceptionOr<> too.

jyavenard replied:
done

cdumez commented on Source/WebCore/html/HTMLMediaElement.cpp:
```diff
@@ -1678,55 +1681,105 @@ void HTMLMediaElement::loadResource(const URL& initialURL, const ContentType& co
 
-    if (m_mediaSource) {
-        loadAttempted = true;
+    auto completionHandler = [url, player = m_player, keySystem, this](const ContentType& contentType) {
```
Capturing `this` here without any kind of protection is fragile, see my other comment.

jyavenard replied:
I'll make it so that it doesn't capture this, but instead receive HTMLMediaElement& as argument, that would alleviate your concerns I think

jyavenard replied:
strike the above.

cdumez commented on Tools/TestWebKitAPI/Tests/WebCore/MIMESniffer.cpp:
```diff
@@ -0,0 +1,52 @@
+
+TEST(MIMESniffer, MIMETypes)
+{
+    static std::span<const uint8_t> oggData = { { 'O', 'g', 'g', 'S', 0, 'm', 'e', 'e', 'e', 'h', '.' } };
```
I am really not convinced this is safe. I believe your span is dangling on the next line since it is constructed from an initializer list which is gone on the next line.

jyavenard replied:
it is not an initialiser list, this is only added in C++26 (https://en.cppreference.com/w/cpp/container/span/span)
I can re-work it if that makes it clear.

cdumez replied:
Ok, it's using the constructor from an array. I think my point stands that it is unsafe / UB:
https://quuxplusone.github.io/blog/2021/10/03/p2447-span-from-initializer-list/

```
std::span<const int> sp = {{1, 2, 3}};  // OK
for (int i : sp) std::cout << i;  // UB, possible segfault
```

cdumez commented on Tools/TestWebKitAPI/Tests/WebCore/MIMESniffer.cpp:
```diff
@@ -0,0 +1,52 @@
+
+#include <WebCore/MIMESniffer.h>
+
+using namespace WebCore;
```
I think it is better practice to put the `using namespace ...` statements *inside* the `namespace TestWebKitAPI {` scope.
I believe we've moved away from using `using namespace ...` statements like this outside namespaces because it causes trouble with our bundled builds (I forget the exact name but you know, how we bundle tons of cpp files together to build faster).

cdumez commented on Source/WebCore/platform/graphics/MediaResourceSniffer.cpp:
```diff
@@ -0,0 +1,102 @@
+    m_content.reset();
+}
+
+Ref<MediaResourceSniffer::Promise> MediaResourceSniffer::promise() const
```
Why does this return a `Ref`. I would expect this to return a `MediaResourceSniffer::Promise&`, unlike a `protectedPromise()` getter which would return a smart pointer, as per our recent smart pointer practices.

jyavenard replied:
promise is a RefCounted object. This is the style I used from the start.
I had never considered to change this style.

cdumez replied:
You seem to misunderstand my comment. I am not asking for the promise to NOT be RefCounted. I am asking you to return a C++ reference.

cdumez replied:
We've had this discussion in the past with `HTMLMediaElement::player()`, which I had to fixup.

cdumez commented on Source/WebCore/platform/graphics/MediaResourceSniffer.cpp:
```diff
@@ -0,0 +1,102 @@
+
+void MediaResourceSniffer::cancel()
+{
+    if (auto resource = WTFMove(m_resource))
```
`std::exchange()` is slightly better than `WTFMove()` here since it guarantees `m_resource` is left in a "good" state (null). Our move constructors often null out things to reduce the risks of bugs but it is still better practice to use std::exchange().

cdumez commented on Source/WebCore/platform/graphics/MediaResourceSniffer.cpp:
```diff
@@ -0,0 +1,102 @@
+    if (!resource)
+        return nullptr;
+    auto sniffer = adoptRef(*new MediaResourceSniffer { Ref { *resource }, maxSize });
+    auto result = sniffer.copyRef();
```
not needed.

cdumez commented on Source/WebCore/platform/graphics/MediaResourceSniffer.cpp:
```diff
@@ -0,0 +1,102 @@
+        return nullptr;
+    auto sniffer = adoptRef(*new MediaResourceSniffer { Ref { *resource }, maxSize });
+    auto result = sniffer.copyRef();
+    resource->setClient(WTFMove(sniffer));
```
`resource->setClient(sniffer.copyRef());`

jyavenard replied:
I've seen that code style in multiple places and I keep copying it. Will change it here and from now on.

cdumez commented on Source/WebCore/platform/graphics/MediaResourceSniffer.cpp:
```diff
@@ -0,0 +1,102 @@
+    auto sniffer = adoptRef(*new MediaResourceSniffer { Ref { *resource }, maxSize });
+    auto result = sniffer.copyRef();
+    resource->setClient(WTFMove(sniffer));
+    return result;
```
`return sniffer;`

cdumez commented on Source/WebCore/html/HTMLMediaElement.cpp:
```diff
@@ -1678,55 +1681,105 @@ void HTMLMediaElement::loadResource(const URL& initialURL, const ContentType& co
-            mediaLoadingFailed(MediaPlayer::NetworkState::FormatError);
+        if (m_mediaSource) {
+            ALWAYS_LOG(LOGIDENTIFIER, "loading MSE blob");
+            if (!m_mediaSource->attachToElement(WeakPtr { *this })) {
```
`WeakPtr { }` is unnecessary.

jyavenard replied:
`virtual bool attachToElement(WeakPtr<HTMLMediaElement>&&) = 0;`

cdumez replied:
Yes, but my comment still stands, you don't need the explicit `WeakPtr { }` as far as I know and you can just pass `*this`.

cdumez commented on Source/WebCore/html/HTMLMediaElement.cpp:
```diff
@@ -1678,55 +1681,105 @@ void HTMLMediaElement::loadResource(const URL& initialURL, const ContentType& co
-    mediaPlayerRenderingModeChanged();
+void HTMLMediaElement::sniffForContentType(const URL& url, Function<bool(const ContentType&)>&& completionHandler)
+{
+    auto handler = [this, completionHandler = WTFMove(completionHandler)](const ContentType& contentType) {
```
Capturing `this` like this without protection / weak ptr is very fragile.

cdumez commented on Source/WebCore/html/HTMLMediaElement.cpp:
```diff
@@ -2667,6 +2720,15 @@ void HTMLMediaElement::mediaLoadingFailed(MediaPlayer::NetworkState error)
         mediaLoadingFailedFatally(error);
+    else if (error == MediaPlayer::NetworkState::FormatError && m_player && hasCurrentSrc() && !m_sniffingAttempted) {
+        // We couldn't find a suitable MediaPlayer, this could be due to the content-type having been initially set incorrectly.
+        sniffForContentType(currentSrc(), [this, url = currentSrc(), player = m_player](const auto& contentType) {
```
Capturing `this` without protection or weak ptr looks unsafe and is fragile at best.

cdumez commented on Source/WebCore/html/HTMLMediaElement.cpp:
```diff
@@ -6237,6 +6299,14 @@ void HTMLMediaElement::cancelPendingTasks()
+
+void HTMLMediaElement::cancelSniffer()
+{
+    if (auto sniffer = WTFMove(m_sniffer))
```
This should ideally use `std::exchange()`.

cdumez commented on Source/WebCore/platform/graphics/MIMESniffer.cpp:
```diff
@@ -0,0 +1,341 @@
+    // Let bitrate-index be sequence[s + 2] & 0xf0 >> 4.
+    auto bitrateIndex = sequence[2] & 0xf0 >> 4;
+    // If the version & 0x01 is non-zero, let bitrate be the value given by bitrate-index in the table mp2.5-rates
+    const uint32_t mp25ratesTable[] = { 0, 8000, 16000, 24000, 32000, 40000, 48000, 56000, 64000, 80000, 96000, 112000, 128000, 144000, 160000 };
```
Would be nice to use `std::array` in new code. Same comment below.
You can also use `constexpr`.

jyavenard replied:
annoying that you have to set the size in the template though, auto-deduction when you want the data to be `const uint8_t` doesn't make for nice syntax

cdumez commented on Source/WebCore/platform/graphics/MIMESniffer.h:
```diff
@@ -0,0 +1,42 @@
+
+namespace WebCore {
+
+class MIMESniffer {
```
Seems like this could be a namespace with a `getMIMETypeFromContent()` free function? We don't seem to need a class at all here.

cdumez commented on Source/WebCore/platform/graphics/MIMESniffer.h:
```diff
@@ -0,0 +1,42 @@
+    WEBCORE_EXPORT static String getMIMETypeFromContent(std::span<const uint8_t>);
+
+private:
+    static String mimeTypeFromSnifferEntries(std::span<const uint8_t>);
```
Could be moved to the cpp if you switch to a namespace.

cdumez commented on Source/WebCore/platform/graphics/MediaPlayer.cpp:
```diff
@@ -208,7 +208,7 @@ class NullMediaPlayerClient : public MediaPlayerClient {
     const Vector<WebCore::ContentType>& mediaContentTypesRequiringHardwareSupport() const final { return nullContentTypeVector(); }
 
-    RefPtr<PlatformMediaResourceLoader> mediaPlayerCreateResourceLoader() final { return nullptr; }
+    RefPtr<PlatformMediaResourceLoader> mediaPlayerCreateResourceLoader() final { ASSERT_NOT_REACHED(); return nullptr; }
```
Surprised we don't we need `NO_RETURN_DUE_TO_ASSERT` here.

jyavenard replied:
this macro doesn't appear to work unless your method returns `void`

cdumez commented on Source/WebCore/platform/graphics/MediaResourceSniffer.cpp:
```diff
@@ -0,0 +1,102 @@
+    auto resource = loader.requestResource(WTFMove(request), PlatformMediaResourceLoader::LoadOption::DisallowCaching);
+    if (!resource)
+        return nullptr;
+    auto sniffer = adoptRef(*new MediaResourceSniffer { Ref { *resource }, maxSize });
```
`Ref { }` is not needed.

cdumez replied:
Also please use `Ref` instead of `auto`. We prefer not to hide smart pointers behind `auto` in new code.

cdumez commented on Source/WebCore/platform/graphics/MediaResourceSniffer.cpp:
```diff
@@ -0,0 +1,102 @@
+    if (m_producer)
+        return;
+    auto buffer = m_content.takeAsContiguous();
+    auto mimeType = MIMESniffer::getMIMETypeFromContent({ buffer->data(), buffer->size() });
```
Should use `dataAsSpanForContiguousData()`

cdumez commented on Source/WebCore/platform/graphics/MediaResourceSniffer.h:
```diff
@@ -0,0 +1,61 @@
+namespace WebCore {
+
+class MediaResourceSniffer final
+    : public PlatformMediaResourceClient {
```
We normally put this on the previous line, except in the case where there are PLENTY of base classes. Here, it's not an issue.

jyavenard committed:
Add ability to sniff media content should we failed to play a file or if the provided content type was invalid.
https://bugs.webkit.org/show_bug.cgi?id=270975
rdar://problem/124614908

Reviewed by Jer Noble.

We have always relied on the server to provide a valid content-type and if that failed
we always used the URL's file name extension instead.
While the HTML5 specs clearly states that determining the type of a resource
should be done through sniffing, this is however a too significant change
to enable right away. There are also performance advantages in using the provided
content-type: it's typically immediately available.
Should the provided content-type be invalid or non-existent (such as with some blobs)
playback would have failed.

We add a MIMEtype sniffer for media content as per HTML5 specs.
Should we fail to play a media using the older content type detection rather
than immediately fail, we will also sniff the type as a last attempt before
retrying.

In order to not unnecessarily retry following a sniffing, we needed to
distinguish a format error from a network error which the MediaPlayerPrivateAVFoundationObjC
didn't do. As a fly-by fix, we add a way to distinguish the two errors by
recording in the WebCoreNSURLSession any network failures.

For now will limit sniffing to media element where the src attribute is set
(and so exclude element where alternative sources are defined)

Added API tests for the mimetype sniffer and tests for webm and mp4 in a blob.

* LayoutTests/media/video-src-mp4-blob-expected.txt: Added.
* LayoutTests/media/video-src-mp4-blob.html: Added.
* LayoutTests/media/video-src-webm-blob-expected.txt: Added.
* LayoutTests/media/video-src-webm-blob.html: Added.
* LayoutTests/media/video-srcobject-mp4-blob-expected.txt: Added.
* LayoutTests/media/video-srcobject-mp4-blob.html: Added.
* LayoutTests/http/tests/media/video-throttled-load-metadata-expected.txt:
* LayoutTests/http/tests/media/video-throttled-load-metadata.html: The test was racy and could
caused the loadedmetadata message to have been received before the worker's message.
* LayoutTests/platform/mac-wk1/TestExpectations:
* LayoutTests/platform/mac/TestExpectations: Remove failure expectations. Tests were failing as an incorrect
content-type was provided. Now that we sniff if the provided content-type was incorrect
files can properly play again.
* Source/WTF/wtf/NativePromise.h:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/html/HTMLMediaElement.cpp: Move all the logic from MediaPlayer related
to guessing the ContentType if missing into the HTMLMediaElement.
If all have failed, we will sniff the content.
(WebCore::HTMLMediaElement::~HTMLMediaElement):
(WebCore::HTMLMediaElement::loadResource): If the content is blocked, signal it as a network error rather than content error.
Otherwise we will attempt to load the content twice (once for sniffing and once for playing) causing test failures.
(WebCore::HTMLMediaElement::needsContentTypeToPlay const): Add methods to determine if sniffing is ultimately required.
(WebCore::HTMLMediaElement::sniffForContentType):
(WebCore::HTMLMediaElement::mediaLoadingFailed): Add one ultimate attempt if we failed to decode and
didn't encounter a network error and didn't attempt to previously sniff content.
(WebCore::HTMLMediaElement::cancelPendingTasks):
(WebCore::HTMLMediaElement::cancelSniffer):
(WebCore::HTMLMediaElement::mediaPlayerEngineFailedToLoad): Save if a network error was encountered.
If so, we won't attempt to sniff later on. This is required as multiple tests are
relying on the network being accessed only once, doing one extra access for sniffing
made the tests fail.
(WebCore::HTMLMediaElement::mediaPlayerEngineFailedToLoad const): Deleted.
* Source/WebCore/html/HTMLMediaElement.h:
* Source/WebCore/platform/CommonAtomStrings.h:
* Source/WebCore/platform/ContentType.cpp:
(WebCore::ContentType::ContentType):
(WebCore::ContentType::fromURL):
* Source/WebCore/platform/ContentType.h:
(WebCore::ContentType::typeWasInferredFromExtension const):
* Source/WebCore/platform/PlatformMediaError.cpp:
(WebCore::convertEnumerationToString):
* Source/WebCore/platform/PlatformMediaError.h:
* Source/WebCore/platform/graphics/MIMESniffer.cpp: Added.
(WebCore::MIMESniffer::span8):
(WebCore::MIMESniffer::hasSignatureForMP4):
(WebCore::MIMESniffer::parseWebMVint):
(WebCore::MIMESniffer::hasSignatureForWebM):
(WebCore::MIMESniffer::matchMP3Header):
(WebCore::MIMESniffer::mp3FrameSize):
(WebCore::MIMESniffer::parseMP3Frame):
(WebCore::MIMESniffer::hasSignatureForMP3WithoutID3):
(WebCore::MIMESniffer::mimeTypeFromSnifferEntries):
(WebCore::MIMESniffer::getMIMETypeFromContent):
* Source/WebCore/platform/graphics/MIMESniffer.h: Added.
* Source/WebCore/platform/graphics/MediaPlayer.cpp:
(WebCore::applicationOctetStream):
(WebCore::MediaPlayer::load):
(WebCore::MediaPlayer::loadWithNextMediaEngine):
* Source/WebCore/platform/graphics/MediaPlayer.h:
(WebCore::MediaPlayerClient::mediaPlayerEngineFailedToLoad):
(WebCore::MediaPlayerClient::mediaPlayerEngineFailedToLoad const): Deleted.
* Source/WebCore/platform/graphics/MediaResourceSniffer.cpp: Added.
(WebCore::MediaResourceSniffer::create):
(WebCore::MediaResourceSniffer::MediaResourceSniffer):
(WebCore::MediaResourceSniffer::~MediaResourceSniffer):
(WebCore::MediaResourceSniffer::cancel):
(WebCore::MediaResourceSniffer::promise const):
(WebCore::MediaResourceSniffer::dataReceived):
(WebCore::MediaResourceSniffer::loadFailed):
(WebCore::MediaResourceSniffer::loadFinished):
* Source/WebCore/platform/graphics/MediaResourceSniffer.h: Added.
* Source/WebCore/platform/graphics/avfoundation/MediaPlayerPrivateAVFoundation.cpp:
(WebCore::MediaPlayerPrivateAVFoundation::updateStates): We needed to be able
to distinguish FormatError vs NetworkError so that we don't unnecessarily retry
playback following sniffing.
* Source/WebCore/platform/graphics/avfoundation/MediaPlayerPrivateAVFoundation.h:
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateAVFoundationObjC.mm:
(WebCore::MediaPlayerPrivateAVFoundationObjC::assetStatus const): Distinguish
load failure due to decoding error vs networking error.
* Source/WebKit/GPUProcess/media/RemoteMediaPlayerProxy.cpp:
(WebKit::RemoteMediaPlayerProxy::mediaPlayerEngineFailedToLoad):
(WebKit::RemoteMediaPlayerProxy::mediaPlayerEngineFailedToLoad const): Deleted.
* Source/WebKit/GPUProcess/media/RemoteMediaPlayerProxy.h:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Tools/TestWebKitAPI/CMakeLists.txt:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebCore/MIMESniffer.cpp: Added.
(TestWebKitAPI::TEST):

Canonical link: https://commits.webkit.org/276258@main
